management-service/src/main/java/pl/pwr/zpi/auth/oauth2/GoogleOAuthLoginSuccessHandler.java

```diff
@@ -25,76 +22,71 @@
 
-        OAuth2AccessToken oAuth2AccessToken = authorizedClient.getAccessToken();
-        OAuth2RefreshToken oAuth2RefreshToken = authorizedClient.getRefreshToken();
-
-        createOrUpdateUser(authentication);
+        OAuth2AuthorizedClient authorizedClient = authorizedClientService.loadAuthorizedClient(registrationId, authentication.getName());
+        DefaultOidcUser oidcUser = (DefaultOidcUser) authentication.getPrincipal();
+        OAuth2RefreshToken refreshToken = authorizedClient.getRefreshToken();
 
-        ResponseCookie authCookie = cookieService.createAuthCookie(oAuth2AccessToken.getTokenValue(), oAuth2AccessToken.getExpiresAt());
-        response.addHeader("Set-Cookie", authCookie.toString());
-        if (oAuth2RefreshToken == null) {
-            throw new RuntimeException("Refresh token is null. Pls contact developers.");
+        if (refreshToken == null) {
+            throw new RuntimeException("Refresh token is null. Please contact the administrator.");
         }
-        ResponseCookie refreshCookie = cookieService.createRefreshCookie(oAuth2RefreshToken.getTokenValue());
-        response.addHeader("Set-Cookie", refreshCookie.toString());
 
-        OAuth2User oAuth2User = (OAuth2User) authentication.getPrincipal();
+        addCookiesToResponse(response, oidcUser.getIdToken().getTokenValue(), oidcUser.getIdToken().getExpiresAt(), refreshToken.getTokenValue());
 
-        List<GrantedAuthority> updatedAuthorities = new ArrayList<>(authentication.getAuthorities());
-        Authentication updatedAuthentication = new OAuth2AuthenticationToken(
-                oAuth2User,
-                updatedAuthorities,
-                registrationId
-        );
+        createOrUpdateUser(oidcUser);
 
-        SecurityContextHolder.getContext().setAuthentication(updatedAuthentication);
+        getRedirectStrategy().sendRedirect(request, response, redirectUri);
+    }
+
+    private void addCookiesToResponse(HttpServletResponse response, String authToken, Instant expiresAt, String refreshToken) {
+        ResponseCookie authCookie = cookieService.createAuthCookie(authToken, expiresAt);
+        ResponseCookie refreshCookie = cookieService.createRefreshCookie(refreshToken);
 
-        String targetUrl = REDIRECT_URI;
-        getRedirectStrategy().sendRedirect(request, response, targetUrl);
+        response.addHeader("Set-Cookie", authCookie.toString());
+        response.addHeader("Set-Cookie", refreshCookie.toString());
     }
 
-    private void createOrUpdateUser(Authentication authentication) {
-        getOAuthUserFromAuthentication(authentication)
+    private void createOrUpdateUser(DefaultOidcUser oidcUser) {
+        Payload payload = googleOauthTokenService.decodeToken(oidcUser.getIdToken().getTokenValue());
+
+        String email = payload.getEmail();
+        String nickname = (String) payload.get("name");
+        String fallbackNickname = nickname != null ? nickname : email.split("@")[0];
```

Comment:
This can be non-unique, if we don't rely on its uniqueness anywhere, then its fine

Reply:
We don't require it to be unique, this is name that every google user can give yourself